towhee/dataframe/dataframe_v2.py

```diff
@@ -13,207 +13,447 @@
             in a column-based manner.
     """
 
     def __init__(
         self,
-        name: str = None,
-        data=None,
-        columns=None,
+        columns: List[Tuple[str, str]],
+        name: str,
     ):
         self._name = name
-        self._len = 0
         self._sealed = False
-        self._lock = threading.Lock()
-
-        # For `data` is empty
-        if not data:
-            pass
-
-        # For `data` is `list`
-        elif isinstance(data, list):
-            container_types = set(type(i) for i in data)
-            if len(container_types) != 1:
-                raise ValueError(
-                    'can not construct Dataframe from a list of hybrid data containers. Try list[Tuple] or list[Array].')
-            container_type = container_types.pop()
-
-            # For `data` is `list[tuple]`
-            if container_type is tuple:
-                self._from_tuples(data, columns)
-            # For `data` is `list[towhee.Array]`
-            elif container_type is Array:
-                self._from_arrays(data, columns)
-            else:
-                raise ValueError('can not construct DataFrame from list[%s]' % (container_type))
 
-        # For `data` is `dict`
-        elif isinstance(data, dict):
-            self._from_dict(data)
+        self._iterator_lock = threading.RLock()
+        self._it_id = 0
+        self._iterators = {}
+        self._min_offset = 0
 
-        # Unrecognized data types
-        else:
-            raise ValueError('can not construct DataFrame from data type %s' % (type(data)))
+        self._block_lock = threading.RLock()
+        self._map_blocked = {}
+        self._window_blocked = {}
+
+
+        self._data_lock = threading.RLock()
+        self._len = 0
+        self._data_as_list = []
+        self._schema = _Schema()
+
+        self._initialize_storage(columns)
+
+    def _initialize_storage(self, columns):
+        """Set the columns in the schema."""
+        for name, tipe in columns:
+            self._schema.add_col(name=name, col_type = tipe)
+        if self._schema.col_index('_frame') is None:
+            self._schema.add_col('_frame', '_Frame')
+        self._data_as_list = [Array(name=name) for name, _ in self._schema.cols]
 
     def __getitem__(self, key):
-        # access a row
-        if isinstance(key, int):
-            return tuple(self._data_as_list[i][key] for i in range(len(self._data_as_list)))
-        # access a column
-        elif isinstance(key, str):
-            return self._data_as_dict[key]
+        """Get data at the passed in offset."""
+        with self._data_lock:
+            # access a row
+            if isinstance(key, int):
+                return tuple(self._data_as_list[i][key] for i in range(len(self._data_as_list)))
+            # access a column
+            elif isinstance(key, str):
+                index = self._schema.col_index(key)
+                return self._data_as_list[index]
+
+    def __str__(self):
```

Comment:
We may only print the metadata, for the main data of dataframe are images.

Reply:
yes, I only included this one now for debugging purposes.